### PR TITLE
WIP: ResolveScopeEnlarger ide dsl

### PR DESCRIPTION
**Pull request by matiaslev**
#465 
please let me know if these make sense and if something more needs to be added in the scope of this ticket.

### CODE REVIEW

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import arrow.meta.phases.ExtensionPhase
+import com.intellij.openapi.extensions.LoadingOrder
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.ResolveScopeEnlarger
+import com.intellij.psi.search.SearchScope
+
+interface ResolveScopeSyntax {
+    fun IdeMetaPlugin.resolveScope(
+        getAdditionalResolveScope: (file: VirtualFile, project: Project) -> SearchScope
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt]:
```suggestion
        additionalResolveScope: (file: VirtualFile, project: Project?) -> SearchScope?
```
Nit's in the type signature

**Comment by i-walker** [comment]
Yes, that's a good angle 👍 
Now the algebra needs extension functions like `ResolveScopeSyntax.searchScope(...): SearchScope` and other smaller functions, which are then composed together in one function like `IdeMetaPlugin.addResolveScopeEnlarger(...): ExtensionPhase`.

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt (hunk 1 above)
Comment (matiaslev) [reply idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt]: thanks!

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    ): ExtensionPhase =
+      extensionProvider(
+        ResolveScopeEnlarger.EP_NAME,
+        object : ResolveScopeEnlarger() {
+            override fun getAdditionalResolveScope(file: VirtualFile, project: Project?): SearchScope? =
+                getAdditionalResolveScope(file, project)
+        },
+        LoadingOrder.FIRST
```
[the remark this answers is not in the document]
Comment (matiaslev) [reply idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt]: thanks! I don't understand the implications of the different LoadingOrder so far haha

[the remark this answers is not in the document]
Comment (i-walker) [reply idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt]: No worries, for the standard behavior of this extension it should not matter.

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt (hunk 1 above)
Comment (i-walker) [reply idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt]: I updated the suggestion from above. But that means to change the signature in the overridden function, too.

File: idea-plugin/src/main/kotlin/arrow/meta/ide/phases/resolve/ResolveScopeProvider.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package arrow.meta.ide.phases.resolve
+
+import arrow.meta.phases.ExtensionPhase
+import com.intellij.psi.search.SearchScope
+
+sealed class ResolveScopeProvider : ExtensionPhase {
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/phases/resolve/ResolveScopeProvider.kt:6]: That file is not needed

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    ): ExtensionPhase =
+      extensionProvider(
+        ResolveScopeEnlarger.EP_NAME,
+        object : ResolveScopeEnlarger() {
+            override fun getAdditionalResolveScope(file: VirtualFile, project: Project): SearchScope =
+              additionalResolveScope(file, project)
+        }
+      )
+
+    fun IdeMetaPlugin.addResolveScopeEnlarger(
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveScopeSyntax.kt:23]: This function is not needed

**Comment by i-walker** [comment]
Looking good :)

In retrospect, there is no need for a function `addResolveScopeEnlarger`, that is built through composition. 
You may address the comments and rename `resolveScope` to `addResolveScopeEnlarger`.

**Comment by i-walker** [comment]
🎉 Lets merge this :) Thanks @matiaslev

**Pull request merged**